File: apps/macos/Shared/VaultConfig.swift
import Foundation

/// The one piece of state shared between `VaultSigner.app` (which knows
/// which vault the user opened) and `VaultSignerAgent` (which needs to
/// know which vault to open when launchd starts it at login, long before
/// any UI exists to tell it) — spec §8's login-item agent has no `--vault`
/// argument to hand it, since a `LaunchAgents` plist's `ProgramArguments`
/// are static and baked in at embed time, not per-user-configurable
/// after install. Neither side stores anything secret here — just the
/// file path, never a passphrase (see `AutoUnlockStore` for that).
enum VaultConfig {
    private static var configPath: String {
        (NSHomeDirectory() as NSString).appendingPathComponent("Library/Application Support/VaultSigner/config.json")
    }

    private static func load() -> [String: Any] {
        guard let data = try? Data(contentsOf: URL(fileURLWithPath: configPath)),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return [:] }
        return json
    }

    private static func write(_ json: [String: Any]) {
        let dir = (configPath as NSString).deletingLastPathComponent
        try? FileManager.default.createDirectory(atPath: dir, withIntermediateDirectories: true, attributes: [.posixPermissions: 0o700])
        let data = try? JSONSerialization.data(withJSONObject: json)
        try? data?.write(to: URL(fileURLWithPath: configPath), options: .atomic)
    }

    static func save(vaultPath: String) {
        var json = load()
        json["vault_path"] = vaultPath
        write(json)
    }

    static func loadVaultPath() -> String? {
        load()["vault_path"] as? String
    }

    /// The one compartment, if any, the agent should unlock automatically
    /// at launch using `AutoUnlockStore` — never the passphrase itself.
    static func saveAutoUnlockCompartmentId(_ compartmentId: String?) {
        var json = load()
        if let compartmentId {
            json["auto_unlock_compartment_id"] = compartmentId
        } else {
            json.removeValue(forKey: "auto_unlock_compartment_id")
        }
        write(json)
    }

    static func loadAutoUnlockCompartmentId() -> String? {
        load()["auto_unlock_compartment_id"] as? String
    }
}

File: apps/windows/VaultSignerAgent/VaultConfig.cs
using System.Text.Json;

namespace VaultSignerAgent;

/// The one piece of state shared between VaultSignerUI (which knows which
/// vault the user opened) and VaultSignerAgent (which needs to know which
/// vault to open when it's started at login, long before any UI exists to
/// tell it) — mirrors apps/macos/Shared/VaultConfig.swift exactly. Neither
/// side stores anything secret here, just the file path — see
/// DpapiAutoUnlockStore for the passphrase itself.
internal static class VaultConfig
{
    private static string ConfigPath =>
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "VaultSigner", "config.json");

    private sealed record ConfigData(string? VaultPath, string? AutoUnlockCompartmentId);

    private static ConfigData Load()
    {
        try
        {
            var json = File.ReadAllText(ConfigPath);
            return JsonSerializer.Deserialize<ConfigData>(json) ?? new ConfigData(null, null);
        }
        catch
        {
            return new ConfigData(null, null);
        }
    }

    private static void Write(ConfigData data)
    {
        var dir = Path.GetDirectoryName(ConfigPath)!;
        Directory.CreateDirectory(dir);
        var tmp = ConfigPath + ".tmp";
        File.WriteAllText(tmp, JsonSerializer.Serialize(data));
        File.Move(tmp, ConfigPath, overwrite: true);
    }

    public static void SaveVaultPath(string vaultPath)
    {
        var current = Load();
        Write(current with { VaultPath = vaultPath });
    }

    public static string? LoadVaultPath() => Load().VaultPath;

    public static void SaveAutoUnlockCompartmentId(string? compartmentId)
    {
        var current = Load();
        Write(current with { AutoUnlockCompartmentId = compartmentId });
    }

    public static string? LoadAutoUnlockCompartmentId() => Load().AutoUnlockCompartmentId;
}
